pkg/operator/target_status_test.go

```diff
@@ -0,0 +1,1532 @@
+	active.LastScrapeDuration = 8.3
+	// We didn't tick yet so we don't expect a change yet.
+	expectStatus(t, "second wait", statusTick2)
+
+	timer.tick()
+	statusTick3 := []v1.ScrapeEndpointStatus{
+		{
+			Name:             "PodMonitoring/gmp-test/prom-example-1/metrics",
+			ActiveTargets:    1,
+			UnhealthyTargets: 0,
+			SampleGroups: []v1.SampleGroup{
+				{
+					SampleTargets: []v1.SampleTarget{
+						{
+							Health: "up",
+							Labels: map[model.LabelName]model.LabelValue{
+								"instance": "a",
+							},
+							LastError:                 lastError("err z"),
+							LastScrapeDurationSeconds: "8.3",
+						},
+					},
+					Count: count(1),
+				},
+			},
+			CollectorsFraction: "1",
+		},
+	}
+	expectStatus(t, "third tick", statusTick3)
+}
+
+// Tests that for pod, targets are fetched correctly (concurrently).
+func TestFetching(t *testing.T) {
+	ctx := context.Background()
+	logger := testr.New(t)
+	concurrency := uint16(4)
+	opts := Options{
+		ProjectID:             "test-proj",
+		Location:              "test-loc",
+		Cluster:               "test-cluster",
+		TargetPollConcurrency: concurrency,
+	}
+	if err := opts.defaultAndValidate(logger); err != nil {
+		t.Fatal("Invalid options:", err)
+	}
+
+	scheme, err := getScheme()
+	if err != nil {
+		t.Fatal("Unable to get scheme")
+	}
+
+	testData := []uint16{0, 1, 2, concurrency - 1, concurrency, concurrency + 1, concurrency * 3}
+	for amount := range testData {
+		t.Run(fmt.Sprintf("fetch-%d-pods", amount), func(t *testing.T) {
+			port := int32(19090)
+			prometheusTargetMap := make(map[string]*prometheusv1.TargetsResult, amount)
+			targetsExpected := make([]*prometheusv1.TargetsResult, 0, amount)
+			kubeClientBuilder := fake.NewClientBuilder().WithScheme(scheme).WithObjects(&appsv1.DaemonSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      NameCollector,
+					Namespace: opts.OperatorNamespace,
+				},
+				Spec: appsv1.DaemonSetSpec{
+					Selector: &metav1.LabelSelector{},
+					Template: corev1.PodTemplateSpec{
+						ObjectMeta: metav1.ObjectMeta{},
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{{
+								Name: "prometheus",
+								Ports: []corev1.ContainerPort{{
+									Name:          "prom-metrics",
+									ContainerPort: port,
+								}},
+							}},
+						},
+					},
+				},
+			})
+			for i := 0; i < amount; i++ {
+				pod := &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      fmt.Sprintf("pod-%d", i),
+						Namespace: opts.OperatorNamespace,
+					},
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{{
+							Name: "prometheus",
+						}},
+					},
+					Status: corev1.PodStatus{
+						Phase: corev1.PodRunning,
+						PodIP: fmt.Sprint(i),
+						ContainerStatuses: []corev1.ContainerStatus{{
+							Name:  "prometheus",
+							Ready: true,
+						}},
+					},
+				}
+				kubeClientBuilder.WithObjects(pod)
+
+				prometheusTargetMap[getPodKey(pod, port)] = &prometheusv1.TargetsResult{
```

Comment:
IIUC we can just initialize the target result once and copy it to the expected array.

Reply:
As per https://github.com/GoogleCloudPlatform/prometheus-engine/pull/301#discussion_r1026558674, fixing this test helped me uncover some issues here. The target result was supposed to be different per-pod, so I fixed that.